packages/api-client-core/src/types.ts

```diff
@@ -21,15 +21,66 @@ export type LimitToKnownKeys<T, U> = {
  */
 export type VariablesOptions = Record<string, VariableOptions>;
 
+/**
+ * Allows detecting an any type, this is rather tricky:
+ * The type constraint 0 extends 1 is not satisfied (0 is not assignable to 1),
+ * so it should be impossible for 0 extends (1 & T) to be satisfied either, since (1 & T) should be even narrower than 1.
+ * However, when T is any, it reduces 0 extends (1 & any) to 0 extends any, which is satisfied.
+ * That's because any is intentionally unsound and acts as both a supertype and subtype of almost every other type.
+ * source: https://stackoverflow.com/questions/49927523/disallow-call-with-any/49928360#49928360
+ */
+
+type IfAny<T, Y, N> = 0 extends 1 & T ? Y : N;
+
+/**
+ * Convert a schema type into the type that a selection of it must extend
+ *
+ * Example Schema:
+ *
+ * {
+ *   foo: boolean;
+ *   bar?: string;
+ *   nested?: {
+ *     count: number
+ *   }
+ * }
+ *
+ * Example available selection:
+ *
+ * {
+ *   foo?: boolean | null | undefined;
+ *   bar?: boolean | null | undefined;
+ *   nested?: {
+ *     count: boolean | null | undefined
+ *   }
+ * }
+ */
+export type AvailableSelection<Schema> = Schema extends string | number | bigint | null | undefined
+  ? boolean | null | undefined
+  : { [key in keyof Schema]?: AvailableSelection<Schema[key]> };
+
+export type AvailableSelection2<Schema> = Schema extends Array<infer T>
```

Comment:
Two issues with the original:
* doesn't handle arrays properly
* converts boolean fields into `boolean | undefined` instead of `boolean | null | undefined` for some reason

Reply:
Why introduce a copy instead of just changing the original? Some backwards compatibility concern?

Reply:
I see I see. To me that suggests that either the react package needs to change and never pass in `any`, or that this change is actually a bad idea and that `any` occurs "naturally" as an input type for good reasons. Can you hunt down if we want to eliminate `any` entirely or not

Reply:
I can definitely dig into that, I'm just not sure how much of a rabbit hole that's going to be. Does it make sense to ship this with the duplication and then follow up with the attempted merge? Otherwise I'm not sure how much more this is going to drag out.

Reply:
We can merge without the deduplication sure! I think its worth resolving before release though.